lib/multiplexer/multiplexer.go

```diff
@@ -246,44 +256,146 @@ func (m *Mux) detectAndForward(conn net.Conn) {
+		SourceAddress:      sourceAddress.String(),
+		DestinationAddress: destinationAddress.String(),
+		ClusterName:        clusterName,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "could not sign jwt token for PROXY line")
+	}
+
+	pl := ProxyLine{
+		Protocol:    TCP4,
+		Source:      sAddr,
+		Destination: dAddr,
+	}
+	err = pl.AddSignature([]byte(signature), signingCert)
+	if err != nil {
+		return nil, trace.Wrap(err, "could not add signature to proxy line")
+	}
+
+	b, err := pl.Bytes()
+	if err != nil {
+		return nil, trace.Wrap(err, "could not get bytes from proxy line")
+	}
+
+	return b, nil
+}
+
+const maxPROXYPasses = 10 // Arbitrary amount that is larger than what realistically we can get
+
+// detect finds out a type of the connection and returns wrapper that support PROXY protocol
+// We pass `casGetter` as function instead of certs themselves, so it could be called only if it's needed.
+func (m *Mux) detect(conn net.Conn) (*Conn, error) {
 	reader := bufio.NewReader(conn)
 
-	// the first attempt is to parse optional proxy
-	// protocol line that is injected by load balancers
-	// before actual protocol traffic flows.
-	// if the first attempt encounters proxy it
-	// goes to the second pass to do protocol detection
+	// Before actual protocol traffic flows, we try to parse optional PROXY protocol headers,
+	// that can be injected by load balancers or our own proxies. There can be multiple PROXY
+	// headers. After they are parsed, last pass does the actual protocol detection itself.
+	// We allow only one unsigned PROXY header from external sources, if it's enabled, and unlimited
+	// amount (within maxPROXYPasses) of signed headers from our own proxies, which take precedence.
 	var proxyLine *ProxyLine
-	for i := 0; i < 2; i++ {
+	for i := 0; i < maxPROXYPasses; i++ {
 		proto, err := detectProto(reader)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
 
 		switch proto {
 		case ProtoProxy:
-			if !enableProxyProtocol {
+			if !m.EnableExternalProxyProtocol {
 				return nil, trace.BadParameter("proxy protocol support is disabled")
 			}
+			// We allow only one unsigned proxy line
 			if proxyLine != nil {
 				return nil, trace.BadParameter("duplicate proxy line")
 			}
 			proxyLine, err = ReadProxyLine(reader)
 			if err != nil {
 				return nil, trace.Wrap(err)
 			}
+			// repeat the cycle to detect the protocol
 		case ProtoProxyV2:
-			if !enableProxyProtocol {
-				return nil, trace.BadParameter("proxy protocol support is disabled")
+			newProxyLine, err := ReadProxyLineV2(reader)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			verified := false
+			// If TLVs are empty we know it can't be signed, so we don't try to verify to avoid unnecessary load
+			if m.CertAuthoritiesGetter != nil && len(newProxyLine.TLVs) > 0 {
+				hostCACerts := [][]byte{}
+				certAuthorities, err := m.CertAuthoritiesGetter.GetCertAuthorities(m.context, types.HostCA, false)
```

Comment:
I wonder if we can get a context related to the connection here, rather than the one attached to the multiplexer ?

Reply:
At this moment it's basically just cancelation propagation from inner levels of multiplexer context, we don't really have any context related to the connection. Unless you have some idea?

Reply:
This is probably fine tbh. It's a shame Go doesn't offer a context for connections out of the box.